### PR TITLE
Fix Dark Mode Flicker and Handle Invalid Locales

### LINKED ISSUE #87: Locale menu should validate current locale
https://beta.elixirschool.com/klingon makes it think like we're supporting that language, but we're not :D

Captura de pantalla 2021-07-23 a las 11 29 38

### PR DESCRIPTION
**Pull request by kinson**
Closes #87 & #88.

# Overview

Updated the dark mode logic to add a blocking js script until the theme is set to avoid a brief flicker of light mode on page load (when dark mode is enabled).

Also updated the set locale plug to check the given locale against the list of valid locales and redirect to the home page if the locale is not valid.

Finally, updated a helper function used by the locale menu to redirect the user to the expected locale when they click on one.

https://user-images.githubusercontent.com/1526888/126853018-46075f3d-529e-4320-b764-f03b13199ecd.mp4

### CODE REVIEW

File: lib/school_house_web/views/html_helpers.ex [hunk 1]
```diff
@@ -14,7 +14,7 @@ defmodule SchoolHouseWeb.HtmlHelpers do
   end
 
   def current_page_locale_path(%{request_path: request_path}, locale) do
-    String.replace(request_path, "/#{current_locale()}/", "/#{locale}/", global: false)
+    String.replace(request_path, "/#{current_locale()}", "/#{locale}", global: false)
```
Comment (kinson) [inline lib/school_house_web/views/html_helpers.ex:17]: @jaimeiniesta I know you just updated this line, let me know if I need to find a different solution to not cause a regression related to your fix in #86 !

Comment (jaimeiniesta) [reply lib/school_house_web/views/html_helpers.ex:17]: @kinson good point on the regression check!

The reason to include that trailing slash is ensuring we're only replacing the locale in its position, and not replace a substring of something that looks like a locale but it's not. Imagine that we add a new section on the site to write about the Essential Elixir Libraries That Every Developer Needs To Know and just like the blog, that URL does not include the locale part, it's just https://beta.elixir.school.com/essential

Then the locale menu would generate again broken links, as this test proves -- you can add it to your branch:

```elixir
test "doesn't break URLs starting with something similar to a locale" do
  assert "/essential" ==
            :get
            |> build_conn("/essential")
            |> HtmlHelpers.current_page_locale_path("ar")
end
```

which fails with:

```
  1) test current_page_locale_path/2 doesn't break URLs starting with something similar to a locale (SchoolHouseWeb.HtmlHelpersTest)
     test/school_house_web/views/html_helpers_test.exs:41
     Assertion with == failed
     code:  assert "/essential" == :get |> build_conn("/essential") |> HtmlHelpers.current_page_locale_path("ar")
     left:  "/essential"
     right: "/arsential"
     stacktrace:
       test/school_house_web/views/html_helpers_test.exs:42: (test)
```

I should have added that test to  protect against this, can you add it to your branch please?

I think we should leave that trailing slash there, if it doesn't break anything else.

**Review by jaimeiniesta — CHANGES_REQUESTED** [review]
It works great, I've tried it on my dev and the issues are gone.

I just left a comment on the locale menu regression.

File: lib/school_house_web/views/html_helpers.ex [hunk 2]
```diff
@@ -14,7 +14,9 @@ defmodule SchoolHouseWeb.HtmlHelpers do
   end
 
   def current_page_locale_path(%{request_path: request_path}, locale) do
-    String.replace(request_path, "/#{current_locale()}/", "/#{locale}/", global: false)
+    request_path
+    |> String.replace(~r/^\/#{current_locale()}(\/|$)/, "/#{locale}/", global: false)
+    |> String.trim_trailing("/")
```
Comment (kinson) [inline lib/school_house_web/views/html_helpers.ex:19]: @jaimeiniesta  thanks for walking through that use case, I had not accounted for a link like `/essential` and did break it! 

The new changes will just check the beginning of the request path for the locale and replace it there if it exists. I two new tests for this as well (including your suggestion, thanks!). This handles the case where the user switches locales from the home page (`/es`) or from a blog post (`/blog/essential-elixir`).

A couple examples of switch from `es` to `ar` for example:
```
/es -> /ar
/blog/server-side-svg-charts-with-contex-and-liveview -> /blog/server-side-svg-charts-with-contex-and-liveview
/es/intermediate/escripts -> /ar/intermediate/escripts
```

**Comment by doomspork** [comment]
Thank you both for working through this to come up with the best solution 🎉 

@jaimeiniesta when you're happy feel free to approve and merge 😁

File: lib/school_house_web/views/html_helpers.ex (hunk 2 above)
Comment (jaimeiniesta) [reply lib/school_house_web/views/html_helpers.ex:19]: That's much better! I like that regexp use.

**Review by jaimeiniesta — APPROVED** [review]
🚀

**Pull request merged**